### ISSUE TITLE
`bundle install --standalone` broken with stdlibs for ruby 3.2

### ISSUE THREAD

**Issue by AlfonsoUceda** [rubygems/rubygems#6273]
### Describe the problem as clearly as you can

When using ruby 3.2 and bundler with `--standalone` option I receive a problem related to a stdlib, e.g. json or io-console.

Running `bundle install --standalone --path=./gems` install gems but doesn't finish because I get a warning like this:
```
json-2.6.3 is built in to Ruby, and can't be cached because your Gemfile doesn't have any sources that contain it.
```
and at the end I see this error:
```
Bundler::GemNotFound: Could not find json-2.6.3.gem for installation
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/source/rubygems.rb:159:in `install'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/installer/gem_installer.rb:54:in `install'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/installer/parallel_installer.rb:155:in `do_install'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/installer/parallel_installer.rb:146:in `block in worker_pool'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/worker.rb:62:in `apply_func'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/worker.rb:57:in `block in process_queue'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/worker.rb:54:in `loop'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/worker.rb:54:in `process_queue'
  /Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/site_ruby/3.2.0/bundler/worker.rb:90:in `block (2 levels) in create_threads'

An error occurred while installing json (2.6.3), and Bundler cannot continue.

In Gemfile:
  hubbado-style was resolved to 1.1.0, which depends on
    rubocop was resolved to 1.43.0, which depends on
      json
```

This problem doesn't happen with ruby 3.1

### Did you try upgrading rubygems & bundler?

I'm using lastest version:
- rubygems 3.4.3
- bundler 2.4.3

### Which command did you run?

`bundle install --standalone --path=./gems`

### What were you expecting to happen?

All gems are installed like in ruby 3.1

### What actually happened?

Couldn't finish `bundle install` process

### If not included with the output of your command, run `bundle env` and paste the output below

```
## Environment

Bundler       2.4.3
  Platforms   ruby, arm64-darwin-21
Ruby          3.2.0p0 (2022-12-25 revision a528908271c678360d2d8ca232c178e7cdd340b4) [arm64-darwin-21]
  Full Path   /Users/alfonsouceda/.rubies/ruby-3.2.0/bin/ruby
  Config Dir  /Users/alfonsouceda/.rubies/ruby-3.2.0/etc
RubyGems      3.4.3
  Gem Home    /Users/alfonsouceda/.gem/ruby/3.2.0
  Gem Path    /Users/alfonsouceda/.gem/ruby/3.2.0:/Users/alfonsouceda/.rubies/ruby-3.2.0/lib/ruby/gems/3.2.0
  User Home   /Users/alfonsouceda
  User Path   /Users/alfonsouceda/.gem/ruby/3.2.0
  Bin Dir     /Users/alfonsouceda/.gem/ruby/3.2.0/bin
Tools
  Git         2.37.1 (Apple Git-137.1)
  RVM         not installed
  rbenv       not installed
  chruby      0.3.9

## Bundler Build Metadata

Built At          2023-01-06
Git SHA           ed1f27f75c
Released Version  true

## Bundler settings

path
  Set for your local app (/Users/alfonsouceda/code/hubbado/components/timesheet_charge_component/.bundle/config): "./gems"
rubygems.pkg.github.com
  Set for the current user (/Users/alfonsouceda/.bundle/config): "alfonsouceda:[REDACTED]"

## Gemfile

### Gemfile

source "https://rubygems.org"
source PRIVATE_HOST

gemspec name: "timesheet_charge_component"

### Gemfile.lock

<No /Users/alfonsouceda/code/hubbado/components/timesheet_charge_component/Gemfile.lock found>

## Gemspecs

### timesheet_charge_component.gemspec

Gem::Specification.new do |s|
  s.name = "timesheet_charge_component"
  s.version = "1.0.0"
  s.authors = PRIVATE_AUTHORS
  s.email = PRIVATE_EMAILS
  s.summary = "Timesheet Charge Component"
  s.description = DESCRIPTION
  s.homepage = PRIVATE_HOMEPAGE

  s.metadata["allowed_push_host"] = PRIVATE_HOST
  s.metadata["github_repo"] = s.homepage
  s.metadata["homepage_uri"] = s.homepage

  s.require_paths = ["lib"]
  s.files = Dir.glob("{lib}/**/*")
  s.platform = Gem::Platform::RUBY
  s.required_ruby_version = ">= 3.0"

  s.add_runtime_dependency "dotenv"
  s.add_runtime_dependency "eventide-postgres"
  s.add_runtime_dependency "evt-component_host"
  s.add_runtime_dependency "hubbado-eventide"
  s.add_runtime_dependency "hubbado-types"
  s.add_runtime_dependency "rollbar"
  s.add_runtime_dependency "charges-client"
  s.add_runtime_dependency "timesheet-client"
  s.add_runtime_dependency "hubbado-idempotence-sequential"

  s.add_development_dependency "evt-messaging-fixtures"
  s.add_development_dependency "evt-entity_projection-fixtures"
  s.add_development_dependency "test_bench"
  s.add_development_dependency "debug"
  s.add_development_dependency "hubbado-style"
end
```

**Comment by deivid-rodriguez** [comment]
Hi for reporting this @AlfonsoUceda.

Can you prepare a repro case for me that does not involve a private source?

**Comment by AlfonsoUceda** [comment]
@deivid-rodriguez I cannot find a reproducible example with a gemspec because I've tried to replicate but doesn't work.

I've been researching bundle code and I've put some `puts`statements here below line 221:
https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/source/rubygems.rb#L221
```
puts "spec.name: #{spec.name}"
puts "cached_path: #{cached_path}"
puts "cached_path.nil?: #{cached_path.nil?}"
puts "remote_specs: #{remote_specs.inspect}"
puts "remote_specs.search(spec): #{remote_specs.search(spec)}"
```

and when `spec.name` is `json` I get the following:
```
spec.name: json
cached_path:
cached_path.nil?: true
remote_specs: #<Bundler::Index:0x2000 sources=[] specs.size=83>
remote_specs.search(spec): []
```
and next time this line is reached again it displays:
```
spec.name: json
cached_path:
cached_path.nil?: true
remote_specs: #<Bundler::Index:0x2000 sources=[] specs.size=84>
remote_specs.search(spec): []
```
It's weird that specs.size increases by one just when this warning happens. Other stdlibs are find, e.g. `io-console` is found before:
```
spec.name: io-console
cached_path:
cached_path.nil?: true
remote_specs: #<Bundler::Index:0x2000 sources=[] specs.size=83>
remote_specs.search(spec): [#<Bundler::EndpointSpecification:0x000000012aa5e120 @extension_dir=nil, @full_gem_path=nil, @gem_dir=nil, @ignored=nil, @bin_dir=nil, @cache_dir=nil, @cache_file=nil, @doc_dir=nil, @ri_dir=nil, @spec_dir=nil, @spec_file=nil, @gems_dir=nil, @base_dir=nil, @loaded=false, @activated=false, @loaded_from=nil, @original_platform=nil, @installed_by_version=nil, @autorequire=nil, @date=nil, @description=nil, @email=nil, @homepage=nil, @name="io-console", @post_install_message=nil, @signing_key=nil, @summary=nil, @version=#<Gem::Version "0.6.0">, @authors=[], @bindir="bin", @cert_chain=[], @dependencies=[], @executables=[], @extensions=[], @extra_rdoc_files=[], @files=[], @licenses=[], @metadata={}, @platform="ruby", @rdoc_options=[], @require_paths=["lib"], @required_ruby_version=#<Gem::Requirement:0x00000001290b9fa8 @requirements=[[">=", #<Gem::Version "2.6.0">]]>, @required_rubygems_version=#<Gem::Requirement:0x00000001290ba138 @requirements=[[">=", #<Gem::Version "0">]]>, @requirements=[], @rubygems_version="3.4.3", @specification_version=4, @test_files=[], @new_platform="ruby", @spec_fetcher=#<Bundler::Fetcher:0x2020 uri=https://rubygems.org/>, @remote_specification=nil, @checksum="ce9d9a3455be48c56cac65b8ef7ec59b07526343905f9fda138791005b563336", @source=#<Bundler::Source::Rubygems:0x2040 rubygems repository https://rubygems.pkg.github.com/hubbado/, https://rubygems.org/ or installed locally>, @remote=#<Bundler::Source::Rubygems::Remote:0x0000000109196f40 @uri=#<Bundler::URI::HTTPS https://rubygems.org/>, @anonymized_uri=#<Bundler::URI::HTTPS https://rubygems.org/>, @cache_slug="rubygems.org.443.29b0360b937aa4d161703e6160654e47">, @full_name="io-console-0.6.0" io-console-0.6.0>]
```

So I have a question, could be because rubygems or bundle don't know how to search `json` because I'm using two sources?

**Comment by deivid-rodriguez** [comment]
Using two global sources like that is deprecated, you should be getting a warning about that, correct? If you change your Gemfile like this, does it work?

```
source "https://rubygems.org"

source PRIVATE_HOST do
  gemspec name: "timesheet_charge_component"
end
```

**Comment by AlfonsoUceda** [comment]
No, it doesn't because tries to fetch all gems from the private source and most of them are public.

**Comment by deivid-rodriguez** [comment]
Interesting, I was expecting it would fallback to the global public source in that case.

Anyways, can you try to isolate whether this is Ruby, RubyGems, or Bundler's fault. For example try downgrading Bundler from 2.4.3 all the way down to 2.3.26 (or whatever version you were using on Ruby 3.1), and see if it starts working again for some older version.

**Comment by AlfonsoUceda** [comment]
@deivid-rodriguez will try to come back with something better, but I guess it is related to the ordering of how dependencies are parsed and fetched from rubygems.

About the versions, I've always had that problem with ruby 3.2 and several versions (2.4.3, 2.4.2, 2.4.1, 2.4.0, 2.3.26) but using 3.1 I don't see that problem using same rubygems and bundler versions.

I have said before I think it is a problem with the sorting and fetching.

One last question, when running `bundle install` does bundler fetches all dependencies at the same order or could change between runs?

**Comment by deivid-rodriguez** [comment]
> About the versions, I've always had that problem with ruby 3.2 and several versions (2.4.3, 2.4.2, 2.4.1, 2.4.0, 2.3.26) but using 3.1 I don't see that problem using same rubygems and bundler versions.

How about RubyGems, did you try downgrading RubyGems `gem update --system 3.3.26` and other previous versions. Also, do you never get the problem in Ruby 3.1, even if you uprade RubyGems and Bundler? That would indeed suggest some issue in Ruby itself.

About ordering, at least if you use `--jobs 1`, I think it should be stable between runs.

**Comment by AlfonsoUceda** [comment]
Yeah, using ruby 3.1 works with rubygems 3.4.3 and bundler 2.4.3.

**Comment by AlfonsoUceda** [comment]
I see this

*Ruby 3.1*
```
*** LOCAL GEMS ***

json (default: 2.6.1)
```

*Ruby 3.2*
```
*** LOCAL GEMS ***

json (default: 2.6.3)
json_schema (0.21.0)
multi_json (1.15.0)
simplecov_json_formatter (0.1.4)
```

**Comment by deivid-rodriguez** [comment]
Can you create a `Gemfile.lock` file with `bundle install` (I assume that works) and copy it here? I don't have any ideas yet, just hope that looking at things lights a bulb in my mind 😅. Also thinking, maybe you can create a dummy GitHub package source and reproduce this against that?

**Comment by AlfonsoUceda** [comment]
This is the Gemfile using bundle install 
```
PATH
  remote: .
  specs:
    timesheet_charge_component (1.0.0)
      charges-client
      dotenv
      eventide-postgres
      evt-component_host
      hubbado-eventide
      hubbado-idempotence-sequential
      hubbado-types
      rollbar
      timesheet-client

GEM
  remote: https://rubygems.org/
  remote: https://rubygems.pkg.github.com/hubbado/
  specs:
    ast (2.4.2)
    charges-client (1.0.0)
      eventide-postgres
      hubbado-idempotence-reservation
      hubbado-types
    concurrent-ruby (1.1.10)
    debug (1.7.1)
      irb (>= 1.5.0)
      reline (>= 0.3.1)
    dotenv (2.8.1)
    dry-core (1.0.0)
      concurrent-ruby (~> 1.0)
      zeitwerk (~> 2.6)
    dry-inflector (1.0.0)
    dry-logic (1.5.0)
      concurrent-ruby (~> 1.0)
      dry-core (~> 1.0, < 2)
      zeitwerk (~> 2.6)
    dry-types (1.7.0)
      concurrent-ruby (~> 1.0)
      dry-core (~> 1.0, < 2)
      dry-inflector (~> 1.0, < 2)
      dry-logic (>= 1.4, < 2)
      zeitwerk (~> 2.6)
    eventide-postgres (2.0.1.0)
      evt-consumer-postgres
      evt-entity_snapshot-postgres
      evt-entity_store
    evt-async_invocation (2.0.0.4)
    evt-attribute (2.0.0.0)
    evt-casing (2.0.0.1)
    evt-clock (2.3.1.1)
      evt-mimic
    evt-component_host (2.0.0.8)
      evt-async_invocation
      evt-casing
      evt-log
      evt-virtual
      ntl-actor
    evt-configure (2.0.0.1)
    evt-consumer (2.3.1.4)
      evt-configure
      evt-messaging
      evt-poll
      evt-settings
      ntl-actor
    evt-consumer-postgres (2.3.2.1)
      evt-consumer
      evt-messaging-postgres
    evt-dependency (2.1.0.0)
      evt-subst_attr
    evt-entity_cache (2.0.1.3)
      evt-configure
      evt-message_store
      evt-settings
      evt-telemetry
    evt-entity_projection (2.0.1.3)
      evt-messaging
    evt-entity_projection-fixtures (1.0.0.1)
      evt-entity_projection
      evt-schema-fixtures
    evt-entity_snapshot-postgres (2.1.1.0)
      evt-entity_store
      evt-messaging-postgres
    evt-entity_store (2.0.4.0)
      evt-entity_cache
      evt-entity_projection
    evt-identifier-uuid (2.1.0.2)
    evt-initializer (2.0.0.0)
      evt-attribute
    evt-invocation (2.2.0.0)
    evt-log (2.1.1.2)
      evt-clock
      evt-dependency
      evt-initializer
      evt-telemetry
      terminal_colors
    evt-message_store (2.4.0.0)
      evt-async_invocation
      evt-casing
      evt-identifier-uuid
      evt-initializer
      evt-schema
      evt-transform
      evt-virtual
    evt-message_store-postgres (2.5.0.0)
      evt-log
      evt-message_store
      evt-settings
      message-db
      pg
    evt-messaging (2.7.0.1)
      evt-message_store
      evt-settings
    evt-messaging-fixtures (1.1.7.0)
      evt-entity_store
      evt-schema-fixtures
    evt-messaging-postgres (2.0.0.0)
      evt-message_store-postgres
      evt-messaging
    evt-mimic (2.3.1.4)
      evt-invocation
    evt-poll (2.0.0.0)
      evt-clock
      evt-log
      evt-telemetry
    evt-reflect (2.2.0.1)
    evt-schema (2.3.1.5)
      evt-attribute
      evt-initializer
      evt-set_attributes
      evt-validate
      evt-virtual
    evt-schema-fixtures (1.0.0.2)
      evt-schema
      test_bench-fixture
    evt-set_attributes (2.2.0.1)
    evt-settings (2.1.1.6)
      evt-attribute
    evt-subst_attr (2.2.0.0)
      evt-attribute
      evt-mimic
      evt-reflect
    evt-telemetry (2.4.0.2)
      evt-clock
      evt-dependency
    evt-transform (2.0.0.0)
      evt-log
      evt-reflect
    evt-try (2.0.0.0)
    evt-validate (2.0.0.0)
      evt-reflect
    evt-virtual (2.1.0.2)
    hubbado-eventide (0.2.3)
      evt-message_store-postgres
    hubbado-idempotence (1.2.0)
      evt-configure
      evt-dependency
      evt-log
      evt-message_store
      evt-messaging-postgres
      evt-try
    hubbado-idempotence-reservation (2.0.0)
      evt-configure
      evt-dependency
      evt-log
      evt-message_store
      evt-messaging-postgres
      evt-try
    hubbado-idempotence-sequential (1.0.1)
      evt-messaging
    hubbado-style (1.1.0)
      rubocop
    hubbado-types (0.2.0)
      dry-types (> 1.0)
    i18n (1.12.0)
      concurrent-ruby (~> 1.0)
    io-console (0.6.0)
    irb (1.6.2)
      reline (>= 0.3.0)
    json (2.6.3)
    message-db (3.0.0)
    money (6.16.0)
      i18n (>= 0.6.4, <= 2)
    ntl-actor (1.4.1)
    parallel (1.22.1)
    parser (3.2.0.0)
      ast (~> 2.4.1)
    pg (1.4.5)
    rainbow (3.1.1)
    regexp_parser (2.6.1)
    reline (0.3.2)
      io-console (~> 0.5)
    rexml (3.2.5)
    rollbar (3.4.0)
    rubocop (1.43.0)
      json (~> 2.3)
      parallel (~> 1.10)
      parser (>= 3.2.0.0)
      rainbow (>= 2.2.2, < 4.0)
      regexp_parser (>= 1.8, < 3.0)
      rexml (>= 3.2.5, < 4.0)
      rubocop-ast (>= 1.24.1, < 2.0)
      ruby-progressbar (~> 1.7)
      unicode-display_width (>= 2.4.0, < 3.0)
    rubocop-ast (1.24.1)
      parser (>= 3.1.1.0)
    ruby-progressbar (1.11.0)
    terminal_colors (1.0.0)
    test_bench (1.2.0.9)
      test_bench-fixture (>= 1.0.0.0)
    test_bench-fixture (1.4.0.2)
    timesheet-client (3.1.0)
      eventide-postgres
      evt-try
      hubbado-eventide
      hubbado-idempotence
      hubbado-types
      money
      rollbar
    unicode-display_width (2.4.2)
    zeitwerk (2.6.6)

PLATFORMS
  arm64-darwin-21

DEPENDENCIES
  debug
  evt-entity_projection-fixtures
  evt-messaging-fixtures
  hubbado-style
  test_bench
  timesheet_charge_component!

BUNDLED WITH
   2.4.3
```

Still trying to have something

**Comment by morkevicius** [comment]
gem update --system 3.3.26 downgrade doesnt solve any problems, i'm getting same thing with different gem withing normal capistrano deployment

**Comment by AlfonsoUceda** [comment]
@morkevicius could you share info about your problem? I mean, is it reproducible with a simple gemfile or gemspec?

**Comment by AlfonsoUceda** [comment]
This is the stacktrace when json gem is not installed:
```
json-2.6.3 is built in to Ruby, and can't be cached because your Gemfile doesn't have any sources that contain it.
Bundler::InstallError: Bundler::GemNotFound: Could not find json-2.6.3.gem for installation
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/source/rubygems.rb:165:in `install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/gem_installer.rb:56:in `install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:156:in `do_install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:140:in `install_serially'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:91:in `call'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer/parallel_installer.rb:67:in `call'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:245:in `install_in_parallel'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:201:in `install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:89:in `block in run'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:12:in `block in lock'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:9:in `open'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/process_lock.rb:9:in `lock'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:71:in `run'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/installer.rb:23:in `install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/cli/install.rb:62:in `run'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:260:in `block in install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/settings.rb:131:in `temporary'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:259:in `install'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:34:in `dispatch'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/cli.rb:28:in `start'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/exe/bundle:45:in `block in <top (required)>'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
  /Users/alfonsouceda/.gem/ruby/3.2.0/gems/bundler-2.4.4/exe/bundle:33:in `<top (required)>'
  /Users/alfonsouceda/.gem/ruby/3.2.0/bin/bundle:25:in `load'
  /Users/alfonsouceda/.gem/ruby/3.2.0/bin/bundle:25:in `<main>'
```

In `lib/bundler/source/rubygems.rb` `install`method, when calling `cached_built_in_gem(spec)` it returns nil, but for io-console, this methods returns the cached path:
```
/Users/alfonsouceda/code/hubbado/components/timesheet_charge_component/gems/ruby/3.2.0/cache/io-console-0.6.0.gem
```

So for some reason, the json gem is not cached in that folder. Still trying to get more info.

**Comment by AlfonsoUceda** [comment]
@deivid-rodriguez is it possible to speak with you in a chat? slack or something like that.

**Comment by deivid-rodriguez** [comment]
You can join the Bundler slack although I'm sometimes slow to respond in there.

**Comment by indirect** [comment]
Here's the minimal reproduction from @composerinteralia in Slack:

```bash
echo 'source "https://rubygems.org/"\ngem "rake"' > Gemfile 
bundle install --standalone

echo "Running with bundler works (it uses the gem method from lib/bundler/rubygems_integration.rb)"
GEM_PATH= GEM_HOME= ruby -r bundler/setup -e 'require "rake"; gem "rake"' 

echo "Running standalone fails (it uses the gem method from lib/rubygems/core_ext/kernel_gem.rb)"
GEM_PATH= GEM_HOME= ruby -r ./bundle/bundler/setup -e 'require "rake"; gem "rake"'
```

**Comment by deivid-rodriguez** [comment]
Does that actually work for any Ruby version? I tried with Ruby 3.1 and Ruby 3.2 and both seem to fail so I'd say it's not related to this issue which seems specific to Ruby 3.2.

In principle that other issue seems expected behavior of the standalone mode for me. My understanding is that people using standalone mode don't want to activate gems like `Kernel.gem` does (discover the gem and put it in the `$LOAD_PATH`) but just want a script that sets up direct requires for whatever libraries they need. If we wanted to implement this feature of "standard Bundler" in the standalone mode, we'd need to port code similar to the following to the generated script:

https://github.com/rubygems/rubygems/blob/451ac56078a3deaacef640906070d3d8f7502570/bundler/lib/bundler.rb#L607-L611

**Comment by indirect** [comment]
@deivid-rodriguez ahhhh I think you are correct. Standalone is supposed to be (as the docs explicitly say) _not_ inclusive of either RubyGems or Bundler at runtime, so the `gem` method is explicitly excluded from that. I don't think we expect or intend the `gem` method to ever work in standalone mode--users should expect no `Gem` constant and no `Bundler` constant at runtime.

**Comment by deivid-rodriguez** [comment]
Yeah! Standalone users definitely expect no `Bundler` constant (I see standalone as a `bundler/setup` replacement), but the `Gem` constant is present anyways, unless you load Ruby with `--disable-gems`.

We could support the above case at the cost of complicating the generated script a bit more, something like this:

```diff
diff --git a/bundler/lib/bundler/installer/standalone.rb b/bundler/lib/bundler/installer/standalone.rb
index 2a8c9a432d..97105cc7fc 100644
--- a/bundler/lib/bundler/installer/standalone.rb
+++ b/bundler/lib/bundler/installer/standalone.rb
@@ -21,6 +21,10 @@ def generate
             file.puts %($:.unshift File.expand_path("\#{__dir__}/#{path}"))
           end
         end
+        if @base_dir
+          file.puts "ENV['GEM_HOME'] = '#{@base_dir}'"
+          file.puts "Gem.clear_paths"
+        end
       end
     end
 
@@ -52,6 +56,9 @@ def bundler_path
 
     def gem_path(path, spec)
       full_path = Pathname.new(path).absolute? ? path : File.join(spec.full_gem_path, path)
+      unless defined?(@base_dir)
+        @base_dir = Pathname.new(path).absolute? ? false : spec.base_dir
+      end
       if spec.source.instance_of?(Source::Path) && spec.source.path.absolute?
         full_path
       else
```

I'm just not sure of the use case.

**Comment by indirect** [comment]
Yeah, my recollection is that we first created standalone before Ruby automatically required 'rubygems', so it didn't even have the Gem constant guaranteed—in the old days, Ruby was like `ruby --disable-gems` by default. So I would hope that we can keep standalone working without `Gem` present.

Anyway I'm also not sure about the usecase, so we can wait to hear about that, potentially.

**Comment by composerinteralia** [comment]
I'm not sure if this matches the original issue here (the Slack thread I had commented in had different details than this issue), but in case it's helpful one problem we had at GitHub is that other libraries outside our control call the `gem` method, and it behaves awkwardly in standalone mode. Assuming bundler isn't loaded at all, a call to `gem` might activate a gem that is not part of our bundle (because it's the `gem` method defined in `lib/rubygems/core_ext/kernel_gem.rb` and not `lib/bundler/rubygems_integration.rb`). And when running with `--disable-gems`, the `gem` method won't be defined at all so we'll get a NoMethodError. We've handled this in our app by defining our own `Kernel#gem` that no-ops.

**Comment by deivid-rodriguez** [comment]
Defining a `Kernel#gem` that no-ops in the generated standalone script sounds nice.

**Comment by deivid-rodriguez** [comment]
@AlfonsoUceda I ended up being able to reproduce your problem, and found a fix at #6918 (unless CI proves me wrong).

Regarding my [previous suggestion](https://github.com/rubygems/rubygems/issues/6273#issuecomment-1381974042) of switching to

```ruby
source "https://rubygems.org"

source PRIVATE_HOST do
  gemspec name: "timesheet_charge_component"
end
```

I also realized why it doesn't actually work as you pointed out. It's because of https://github.com/rubygems/rubygems/issues/4817.

I'll try to implement a fix for that so that you can switch to a secure Gemfile and you are not left behind when Bundler 3 is released.

Regarding the other issue related to standalone mode reported in this thread, I'll open a separate ticket 👍.

**Comment by AlfonsoUceda** [comment]
Thank you so much @deivid-rodriguez will try again when this is released ;)
